component/file_cache/file_cache.go

```diff
@@ -704,107 +739,171 @@ func (fc *FileCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Hand
 	defer flock.Unlock()
```

Comment:
This flock is per-file (not just per-handle), so once it locks, we know that we're the only ones working on this file.
This could be a good opportunity to address the potential issue of two handles trying to download the same file.
Maybe you could call `isDownloadRequired` (again) here, and if it comes back false, that means another handle has already downloaded the file. In that case, you could skip downloading it and just open the already-downloaded file.